File: Cpp 02/rational.cpp
#include "rational.h"

int rational::gcd( int n1, int n2 ){
   if(n1 == 0){
      return n2;
   }if(n2 == 0){
      return n1;
   }
   return gcd(n2%n1, n1);
};

void rational::normalize(){
   int gcd1 = rational::gcd(num, denum);
   if(gcd1 < 0) gcd1 = gcd1 * (-1);
   if(denum < 0){
      denum = denum * (-1);
      num = num * (-1);
   }
   if(gcd1 !=1){
      num = num / gcd1;
      denum = denum / gcd1;
   }
};

rational operator - ( rational r ){
   r.num = r.num * (-1);
   return r;
};

rational operator + ( const rational& r1, const rational& r2 ){
   rational temp;
   if(r1.denum == r2.denum){
      temp.num = r1.num + r2.num;
      temp.denum = r1.denum;
      temp.normalize();
      return temp;
   }else{
      temp.num = r1.num*r2.denum + r2.num*r1.denum;
      temp.denum = r1.denum * r2.denum;
      temp.normalize();
      return temp;
   }
};

rational operator - ( const rational& r1, const rational& r2 ){
   rational temp;
   if(r1.denum == r2.denum){
      temp.num = r1.num - r2.num;
      temp.denum = r1.denum;
      temp.normalize();
      return temp;
   }else{
      temp.num = r1.num*r2.denum - r2.num*r1.denum;
      temp.denum = r1.denum * r2.denum;
      temp.normalize();
      return temp;
   }
};

rational operator * ( const rational& r1, const rational& r2 ){
   rational temp;
   temp.num = r1.num * r2.num;
   temp.denum = r1.denum * r2.denum;
   temp.normalize();
   return temp;
};

rational operator / ( const rational& r1, const rational& r2 ){
   rational temp;
   temp.num = r1.num * r2.denum;
   temp.denum = r1.denum * r2.num;
   temp.normalize();
   return temp;
};

bool operator == ( const rational& r1, const rational& r2 ){
   if(r1.num == r2.num && r1.denum == r2.denum){
      return true;
   }else{
      return false;
   }
};

bool operator != ( const rational& r1, const rational& r2 ){
   if(r1.num != r2.num && r1.denum != r2.denum){
      return true;
   }else{
      return false;
   }
};

std::ostream& operator << ( std::ostream& out, const rational& r ){
   if(r.denum == 1){
   std::cout << r.num;
   }else{
   std::cout << r.num << "/" << r.denum;
   }
};

File: Python 03/rational.py
from numbers import *

def gcd(n1, n2):
    if n1 == 0:
        return n2
    elif n2 == 0:
        return n1
    elif n1 == 0 and n2 == 0:
        raise ArithmeticError("gcd(0,0) does not exist.")
    else:
        return gcd(n2 % n1, n1)

class Rational(Number):
    def __init__(self, num, denum = 1):
        self.num = num
        self.denum = denum
        self.normalize()

    def __repr__(self):
        if self.denum == 1:
            return '{}'.format(self.num)
        else:
            return '{}/{}'.format(self.num, self.denum)

    def normalize(self):
        gc = gcd(self.num, self.denum)
        if gc < 0:
            gc *= -1
        if self.denum < 0:
            self.denum *= -1
            self.num *= -1
        if gc != 1:
            self.denum //= gc
            self.num //= gc

    def __neg__(self):
        return Rational(self.num * (-1), self.denum)

    def __add__(self, other):
        if not isinstance(other, Rational):
            return Rational(self.num + self.denum * other, self.denum)
        else:
            return Rational(self.num * other.denum + other.num * self.denum, self.denum * other.denum)

    def __sub__(self, other):
        if not isinstance(other, Rational):
            return Rational(self.num - self.denum * other, self.denum)
        else:
            return Rational(self.num * other.denum - other.num * self.denum, self.denum * other.denum)

    def __radd_(self, other):
        return self + other;

    def __rsub_(self, other):
        return -(self-other)

    def __mul__(self, other):
        if not isinstance(other, Rational):
            return Rational(self.num * other, self.denum)
        else:
            return Rational(self.num * other.num, self.denum * other.denum)

    def __truediv__(self, other):
        if not isinstance(other, Rational):
            return Rational(self.num, self.denum * other)
        else:
            return Rational(self.num * other.denum, self.denum * other.num)

    def __rmul__(self, other):
        return self * other

    def __rtruediv__(self, other):
        if not isinstance(other, Rational):
            return Rational(self.denum * other, self.num)
